In [ ]:
from pyspark.sql.functions import isnan, when, count, col
import pyspark
from pyspark.ml import Pipeline
from pyspark.ml.feature import StringIndexer, VectorAssembler
from pyspark.ml.classification import LogisticRegression

1. Sample code number            id number
   2. Clump Thickness               1 - 10
   3. Uniformity of Cell Size       1 - 10
   4. Uniformity of Cell Shape      1 - 10
   5. Marginal Adhesion             1 - 10
   6. Single Epithelial Cell Size   1 - 10
   7. Bare Nuclei                   1 - 10
   8. Bland Chromatin               1 - 10
   9. Normal Nucleoli               1 - 10
  10. Mitoses                       1 - 10
  11. Class:                        (2 for benign, 4 for malignant)

In [ ]:
from pyspark.sql.types import DoubleType, StringType, StructField, StructType, LongType

schema = StructType([
  StructField("id", StringType(), False),
  StructField("clump_thickness", DoubleType(), False),
  StructField("uniformity_cell_size", DoubleType(), False),
  StructField("uniformity_cell_shape", DoubleType(), False),
  StructField("marginal_adhesion", DoubleType(), False),
  StructField("single_epithelial_cell_size", DoubleType(), False),
  StructField("bare_nuclei", DoubleType(), False),
  StructField("bland_chromatin", DoubleType(), False),
  StructField("normal_nucleoli", DoubleType(), False),
  StructField("mitoses", DoubleType(), False),
  StructField("cancer", StringType(), False),
])

In [ ]:
from pyspark.sql import SparkSession
spark = SparkSession.builder \
    .master("local") \
    .appName("Word Count") \
    .config("spark.some.config.option", "some-value") \
    .getOrCreate()

In [ ]:
# %fs ls /FileStore/tables/breast_cancer_wisconsin.csv

path name size modificationTime dbfs:/FileStore/tables/breast_cancer_wisconsin.csv breast_cancer_wisconsin.csv 19889 1652709929000

In [ ]:
df_cancer = spark.read.format("csv").schema(schema).load("dbfs:/FileStore/tables/breast_cancer_wisconsin.csv")

In [ ]:
df_cancer.head()

Out[390]: Row(id='1000025', clump_thickness=5.0, uniformity_cell_size=1.0, uniformity_cell_shape=1.0, marginal_adhesion=1.0, single_epithelial_cell_size=2.0, bare_nuclei=1.0, bland_chromatin=3.0, normal_nucleoli=1.0, mitoses=1.0, cancer='2')

In [ ]:
df_cancer = df_cancer.fillna(0)

In [ ]:
df_cancer.toPandas().head()

id 
 clump_thickness 
 uniformity_cell_size 
 uniformity_cell_shape 
 marginal_adhesion 
 single_epithelial_cell_size 
 bare_nuclei 
 bland_chromatin 
 normal_nucleoli 
 mitoses 
 cancer 
 
 
 
 
 0 
 1000025 
 5.0 
 1.0 
 1.0 
 1.0 
 2.0 
 1.0 
 3.0 
 1.0 
 1.0 
 2 
 
 
 1 
 1002945 
 5.0 
 4.0 
 4.0 
 5.0 
 7.0 
 10.0 
 3.0 
 2.0 
 1.0 
 2 
 
 
 2 
 1015425 
 3.0 
 1.0 
 1.0 
 1.0 
 2.0 
 2.0 
 3.0 
 1.0 
 1.0 
 2 
 
 
 3 
 1016277 
 6.0 
 8.0 
 8.0 
 1.0 
 3.0 
 4.0 
 3.0 
 7.0 
 1.0 
 2 
 
 
 4 
 1017023 
 4.0 
 1.0 
 1.0 
 3.0 
 2.0 
 1.0 
 3.0 
 1.0 
 1.0 
 2

In [ ]:
df_cancer.count()

Out[393]: 699

In [ ]:
#df_cancer = df_cancer.replace({2.0: 0.0, 4.0: 1.0 }, subset = ["label"])
df_cancer.toPandas().head()

id 
 clump_thickness 
 uniformity_cell_size 
 uniformity_cell_shape 
 marginal_adhesion 
 single_epithelial_cell_size 
 bare_nuclei 
 bland_chromatin 
 normal_nucleoli 
 mitoses 
 cancer 
 
 
 
 
 0 
 1000025 
 5.0 
 1.0 
 1.0 
 1.0 
 2.0 
 1.0 
 3.0 
 1.0 
 1.0 
 2 
 
 
 1 
 1002945 
 5.0 
 4.0 
 4.0 
 5.0 
 7.0 
 10.0 
 3.0 
 2.0 
 1.0 
 2 
 
 
 2 
 1015425 
 3.0 
 1.0 
 1.0 
 1.0 
 2.0 
 2.0 
 3.0 
 1.0 
 1.0 
 2 
 
 
 3 
 1016277 
 6.0 
 8.0 
 8.0 
 1.0 
 3.0 
 4.0 
 3.0 
 7.0 
 1.0 
 2 
 
 
 4 
 1017023 
 4.0 
 1.0 
 1.0 
 3.0 
 2.0 
 1.0 
 3.0 
 1.0 
 1.0 
 2

In [ ]:
# Number of positive examples
df_cancer[df_cancer["cancer"]=='2'].count()

Out[395]: 458

In [ ]:
# Number of negative examples
df_cancer[df_cancer["cancer"]=='4'].count()

Out[396]: 241

In [ ]:
#df_cancer.filter(df_cancer["label"]==0.0).count()

In [ ]:
df_cancer.select([count(when(isnan(c), c)).alias(c) for c in df_cancer.columns]).show()

+---+---------------+--------------------+---------------------+-----------------+---------------------------+-----------+---------------+---------------+-------+------+
| id|clump_thickness|uniformity_cell_size|uniformity_cell_shape|marginal_adhesion|single_epithelial_cell_size|bare_nuclei|bland_chromatin|normal_nucleoli|mitoses|cancer|
+---+---------------+--------------------+---------------------+-----------------+---------------------------+-----------+---------------+---------------+-------+------+
|  0|              0|                   0|                    0|                0|                          0|          0|              0|              0|      0|     0|
+---+---------------+--------------------+---------------------+-----------------+---------------------------+-----------+---------------+---------------+-------+------+



In [ ]:
# Convert label into label indices using the StringIndexer
label_stringIdx = StringIndexer(inputCol="cancer", outputCol="label")
stages = [label_stringIdx]

In [ ]:
label_stringIdx.getOutputCol()

Out[400]: 'label'

In [ ]:
assemblerInputs = df_cancer.columns[1:10]

In [ ]:
assembler = VectorAssembler(inputCols=assemblerInputs, outputCol="features")

In [ ]:
assemblerInputs

Out[403]: ['clump_thickness',
 'uniformity_cell_size',
 'uniformity_cell_shape',
 'marginal_adhesion',
 'single_epithelial_cell_size',
 'bare_nuclei',
 'bland_chromatin',
 'normal_nucleoli',
 'mitoses']

In [ ]:
df_cancer.columns

Out[404]: ['id',
 'clump_thickness',
 'uniformity_cell_size',
 'uniformity_cell_shape',
 'marginal_adhesion',
 'single_epithelial_cell_size',
 'bare_nuclei',
 'bland_chromatin',
 'normal_nucleoli',
 'mitoses',
 'cancer']

In [ ]:
stages += [assembler]

In [ ]:
stages

Out[406]: [StringIndexer_1d7cf6cf85a1, VectorAssembler_334e76a1761e]

In [ ]:
partialPipeline = Pipeline().setStages(stages)
pipelineModel = partialPipeline.fit(df_cancer)
preppedDataDF = pipelineModel.transform(df_cancer)

In [ ]:
preppedDataDF.toPandas().head()

/databricks/spark/python/pyspark/sql/pandas/conversion.py:92: UserWarning: toPandas attempted Arrow optimization because 'spark.sql.execution.arrow.pyspark.enabled' is set to true; however, failed by the reason below:
  Unable to convert the field features. If this column is not necessary, you may consider dropping it or converting to primitive type before the conversion.
Direct cause: Unsupported type in conversion to Arrow: VectorUDT
Attempting non-optimization as 'spark.sql.execution.arrow.pyspark.fallback.enabled' is set to true.
  warnings.warn(msg)


id 
 clump_thickness 
 uniformity_cell_size 
 uniformity_cell_shape 
 marginal_adhesion 
 single_epithelial_cell_size 
 bare_nuclei 
 bland_chromatin 
 normal_nucleoli 
 mitoses 
 cancer 
 label 
 features 
 
 
 
 
 0 
 1000025 
 5.0 
 1.0 
 1.0 
 1.0 
 2.0 
 1.0 
 3.0 
 1.0 
 1.0 
 2 
 0.0 
 [5.0, 1.0, 1.0, 1.0, 2.0, 1.0, 3.0, 1.0, 1.0] 
 
 
 1 
 1002945 
 5.0 
 4.0 
 4.0 
 5.0 
 7.0 
 10.0 
 3.0 
 2.0 
 1.0 
 2 
 0.0 
 [5.0, 4.0, 4.0, 5.0, 7.0, 10.0, 3.0, 2.0, 1.0] 
 
 
 2 
 1015425 
 3.0 
 1.0 
 1.0 
 1.0 
 2.0 
 2.0 
 3.0 
 1.0 
 1.0 
 2 
 0.0 
 [3.0, 1.0, 1.0, 1.0, 2.0, 2.0, 3.0, 1.0, 1.0] 
 
 
 3 
 1016277 
 6.0 
 8.0 
 8.0 
 1.0 
 3.0 
 4.0 
 3.0 
 7.0 
 1.0 
 2 
 0.0 
 [6.0, 8.0, 8.0, 1.0, 3.0, 4.0, 3.0, 7.0, 1.0] 
 
 
 4 
 1017023 
 4.0 
 1.0 
 1.0 
 3.0 
 2.0 
 1.0 
 3.0 
 1.0 
 1.0 
 2 
 0.0 
 [4.0, 1.0, 1.0, 3.0, 2.0, 1.0, 3.0, 1.0, 1.0]

In [ ]:
# Fit model to prepped data
lrModel = LogisticRegression().fit(preppedDataDF)

# ROC for training data
#display(lrModel, preppedDataDF, "ROC")

In [ ]:
display(lrModel, preppedDataDF, "ROC")

False Positive Rate True Positive Rate Threshold 0.0 0.0 0.9999997760989063 0.0 0.02631578947368421 0.9999997760989063 0.0 0.05263157894736842 0.9999995681991343 0.0 0.07894736842105263 0.9999940519518831 0.0 0.10526315789473684 0.9999935717171475 0.0 0.13157894736842105 0.9999873537710642 0.0 0.15789473684210525 0.9999870512590772 0.0 0.18421052631578946 0.9999424942024002 0.0 0.21052631578947367 0.9999306438461848 0.0 0.23684210526315788 0.9998257613825956 0.0 0.2631578947368421 0.9996619532927052 0.0 0.2894736842105263 0.9995915706702989 0.0 0.3157894736842105 0.9994465622410118 0.0 0.34210526315789475 0.9993111402974599 0.0 0.3684210526315789 0.999311114429155 0.0 0.39473684210526316 0.9991450240314043 0.0 0.42105263157894735 0.9991349397256815 0.0 0.4473684210526316 0.9990134268628967 0.0 0.47368421052631576 0.9989619252304466 0.0 0.5 0.9987502582876545 0.0 0.5263157894736842 0.9979206307854671 0.0 0.5526315789473685 0.9967233235729869 0.0 0.5789473684210527 0.9923347152776113 0.0 0.6052631578947368 0.9915669303101285 0.0 0.631578947368421 0.9893100233232016 0.0 0.6578947368421053 0.989258745344989 0.0 0.6842105263157895 0.9872087729622229 0.0 0.7105263157894737 0.9858063952020591 0.0 0.7368421052631579 0.9851276833180266 0.0 0.7631578947368421 0.9847979735956234 0.0 0.7894736842105263 0.9843628319043766 0.0 0.8157894736842105 0.9818577555562978 0.016129032258064516 0.8157894736842105 0.9781276614373275 0.016129032258064516 0.8421052631578947 0.9727635361012944 0.016129032258064516 0.868421052631579 0.9614355845765654 0.016129032258064516 0.8947368421052632 0.9388363484662495 0.016129032258064516 0.9210526315789473 0.921470853952443 0.03225806451612903 0.9210526315789473 0.9061514373438179 0.03225806451612903 0.9473684210526315 0.7042247947712305 0.03225806451612903 0.9736842105263158 0.5363960932693937 0.03225806451612903 1.0 0.20631916713404352 0.04838709677419355 1.0 0.13881615644252376 0.06451612903225806 1.0 0.09264883945489633 0.08064516129032258 1.0 0.05453332795348931 0.0967741935483871 1.0 0.04631247420207485 0.11290322580645161 1.0 0.032454740157833006 0.12903225806451613 1.0 0.030606639468189756 0.14516129032258066 1.0 0.02358492457045247 0.16129032258064516 1.0 0.02275314420180774 0.1774193548387097 1.0 0.021352773880914833 0.1935483870967742 1.0 0.0204092402964392 0.20967741935483872 1.0 0.01504764675798842 0.22580645161290322 1.0 0.01410593547959016 0.24193548387096775 1.0 0.013104104397741678 0.2903225806451613 1.0 0.012001140396626364 0.3064516129032258 1.0 0.010519598651834494 0.3225806451612903 1.0 0.009850328301759137 0.3387096774193548 1.0 0.008032526081428348 0.3870967741935484 1.0 0.007932675316591253 0.4032258064516129 1.0 0.006996866162881532 0.41935483870967744 1.0 0.006640157286560533 0.45161290322580644 1.0 0.004651485931471799 0.46774193548387094 1.0 0.004407768687775815 0.5806451612903226 1.0 0.004131844137166829 0.5967741935483871 1.0 0.004088363956531116 0.6451612903225806 1.0 0.0036700401940359395 0.6612903225806451 1.0 0.0027237673786997263 0.6774193548387096 1.0 0.0024781238779807424 0.7419354838709677 1.0 0.002418957724638404 0.7580645161290323 1.0 0.002148185042772699 0.7741935483870968 1.0 0.0020348815224680017 0.7903225806451613 1.0 0.0019581789115830084 0.8064516129032258 1.0 0.0015007487250998652 0.8548387096774194 1.0 0.001415151933493508 0.8709677419354839 1.0 0.0013404594538029002 0.8870967741935484 1.0 0.0013055964240799023 0.967741935483871 1.0 9.32020783639409E-4 0.9838709677419355 1.0 9.221817710494129E-4 1.0 1.0 8.828056823420027E-4 1.0 1.0 8.828056823420027E-4

In [ ]:
display(lrModel, preppedDataDF)

fitted values residuals 2.26752399387699 -0.9061514373438179 -4.321639064350649 -0.013104104397741678 -6.195280726993061 -0.0020348815224680017 -6.021996635756776 -0.002418957724638404 -3.4554535128316672 -0.030606639468189756 -4.410679865174761 -0.012001140396626364 -6.021996635756776 -0.002418957724638404 -6.021996635756776 -0.002418957724638404 -1.8251575627490524 -0.13881615644252376 7.063581854659081 8.549759685957392E-4 4.14234391003 0.015637168095623433 -3.7600359447501206 -0.02275314420180774 7.279783835369973 6.888597025400678E-4 -5.603875873607709 -0.0036700401940359395 4.171007782143155 0.015202026404376623 11.254499051756579 1.2948740922835711E-5 4.346121991557374 0.012791227037777131 4.193269716473758 0.014872316681973419 0.14584232976702438 0.46360390673060625 -2.281713115737216 -0.09264883945489633 2.4625013045837445 0.07852914604755701 7.802783145378421 4.084293297010877E-4 -6.5591022259507294 -0.001415151933493508 4.863358807555803 0.007665284722388699 3.2160975970084538 0.038564415423434584 9.763567279970419 5.750579759977725E-5 -6.977222988099972 -9.32020783639409E-4 4.522864017583295 0.010741254655010968 6.683567855851715 0.0012497417123454868 -5.603875873607709 -0.0036700401940359395 -6.639788863209921 -0.0013055964240799023 3.8004174355477027 -0.9781276614373275 -5.484891045562801 -0.004131844137166829 -2.852866601603743 -0.05453332795348931 11.954796679569654 6.428282852466438E-6 6.920286037163182 9.865731371032638E-4 -6.987845597953052 -9.221817710494129E-4 -3.824989581149413 -0.021352773880914833 15.312061204836407 2.2390109366554611E-7 4.76712560527037 0.008433069689871453 7.498807671062178 5.534377589881956E-4 6.869348850565242 0.001038074769553421 -5.495513655415962 -0.004088363956531116 3.5755843670288265 0.0272364638987056 -6.14098146380169 -0.002148185042772699 5.717643598266139 0.0032766764270131254 -3.87114713830501 -0.0204092402964392 -5.603875873607709 -0.0036700401940359395 -4.181371650285292 -0.01504764675798842 -4.543939902609614 -0.010519598651834494 -3.0249247306867337 -0.04631247420207485 -7.031522251291313 -8.828056823420027E-4 -4.610351330478742 -0.009850328301759137 7.279746257832678 6.888855708450103E-4 7.051845937512032 8.650602743185409E-4 -6.5591022259507294 -0.001415151933493508 -3.394915695784166 -0.032454740157833006 12.032441491734637 5.948048116932725E-6 -5.365906217517878 -0.004651485931471799 -4.828800627323856 -0.007932675316591253 -6.977222988099972 -9.32020783639409E-4 7.991988380907672 3.3804670729475905E-4 0.8674978919364449 0.29577520522876954 8.654910580019285 1.742386174043986E-4 6.173609159401718 0.002079369214532889 -4.828800627323856 -0.007932675316591253 14.655300882133856 4.318008657300254E-7 -1.3472570819165592 0.7936808328659565 -6.977222988099972 -9.32020783639409E-4 -4.828800627323856 -0.007932675316591253 -5.365906217517878 -0.004651485931471799 -6.021996635756776 -0.002418957724638404 -5.9030118077118585 -0.0027237673786997263 -5.484891045562801 -0.004131844137166829 -4.410679865174761 -0.012001140396626364 -5.484891045562801 -0.004131844137166829 9.576186320961773 6.935615381520321E-5 -4.410679865174761 -0.012001140396626364 11.278138848765153 1.2646228935819792E-5 -6.613401489142292 -0.0013404594538029002 4.527701210753532 0.010689976676798385 -5.997772306401394 -0.0024781238779807424 -5.484891045562801 -0.004131844137166829 4.240668484752627 0.01419360479794085 -5.484891045562801 -0.004131844137166829 -5.484891045562801 -0.004131844137166829 -6.500289269364928 -0.0015007487250998652 -5.484891045562801 -0.004131844137166829 -4.246953244389807 -0.01410593547959016 -6.5591022259507294 -0.001415151933493508 -5.007957326827837 -0.006640157286560533 -3.723280059677565 -0.02358492457045247 3.9912032800733046 0.01814224444370216 -4.816191259368524 -0.008032526081428348 -6.2337802657384325 -0.0019581789115830084 -6.977222988099972 -9.32020783639409E-4 -6.977222988099972 -9.32020783639409E-4 2.7310880973503173 0.061163651533750474 -4.955271462173965 -0.00699686

In [ ]:
### Randomly split data into training and test sets. set seed for reproducibility
(trainingData, testData) = preppedDataDF.randomSplit([0.7, 0.3], seed=100)
print(trainingData.count())
print(testData.count())

496
203


## Logistic Regression

In [ ]:
# Create initial LogisticRegression model
lr = LogisticRegression(labelCol="label", featuresCol="features", maxIter=10)

# Train model with Training Data
lrModel = lr.fit(trainingData)

In [ ]:
# Make predictions on test data using the transform() method.
# LogisticRegression.transform() will only use the 'features' column.
predictions = lrModel.transform(testData)

In [ ]:
predictions.toPandas().head(5)

/databricks/spark/python/pyspark/sql/pandas/conversion.py:92: UserWarning: toPandas attempted Arrow optimization because 'spark.sql.execution.arrow.pyspark.enabled' is set to true; however, failed by the reason below:
  Unable to convert the field features. If this column is not necessary, you may consider dropping it or converting to primitive type before the conversion.
Direct cause: Unsupported type in conversion to Arrow: VectorUDT
Attempting non-optimization as 'spark.sql.execution.arrow.pyspark.fallback.enabled' is set to true.
  warnings.warn(msg)


id 
 clump_thickness 
 uniformity_cell_size 
 uniformity_cell_shape 
 marginal_adhesion 
 single_epithelial_cell_size 
 bare_nuclei 
 bland_chromatin 
 normal_nucleoli 
 mitoses 
 cancer 
 label 
 features 
 rawPrediction 
 probability 
 prediction 
 
 
 
 
 0 
 1002945 
 5.0 
 4.0 
 4.0 
 5.0 
 7.0 
 10.0 
 3.0 
 2.0 
 1.0 
 2 
 0.0 
 [5.0, 4.0, 4.0, 5.0, 7.0, 10.0, 3.0, 2.0, 1.0] 
 [-2.5580705431065436, 2.5580705431065436] 
 [0.07188616653442371, 0.9281138334655763] 
 1.0 
 
 
 1 
 1018099 
 1.0 
 1.0 
 1.0 
 1.0 
 2.0 
 10.0 
 3.0 
 1.0 
 1.0 
 2 
 0.0 
 [1.0, 1.0, 1.0, 1.0, 2.0, 10.0, 3.0, 1.0, 1.0] 
 [1.8859352405161376, -1.8859352405161376] 
 [0.8682913742058727, 0.13170862579412734] 
 0.0 
 
 
 2 
 1018561 
 2.0 
 1.0 
 2.0 
 1.0 
 2.0 
 1.0 
 3.0 
 1.0 
 1.0 
 2 
 0.0 
 [2.0, 1.0, 2.0, 1.0, 2.0, 1.0, 3.0, 1.0, 1.0] 
 [5.227536157072055, -5.227536157072055] 
 [0.9946619162530372, 0.005338083746962763] 
 0.0 
 
 
 3 
 1031608 
 2.0 
 1.0 
 1.0 
 1.0 
 1.0 
 1.0 
 2.0 
 1.0 
 1.0 
 2 
 0.0 
 [2.0, 1.0, 1.0, 1.0, 1.0, 1.0, 2.0, 1.0, 1.0] 
 [5.956516924091725, -5.956516924091725] 
 [0.9974177714977072, 0.002582228502292838] 
 0.0 
 
 
 4 
 1033078 
 2.0 
 1.0 
 1.0 
 1.0 
 2.0 
 1.0 
 1.0 
 1.0 
 5.0 
 2 
 0.0 
 [2.0, 1.0, 1.0, 1.0, 2.0, 1.0, 1.0, 1.0, 5.0] 
 [4.333165343034546, -4.333165343034546] 
 [0.9870441243816102, 0.012955875618389845] 
 0.0

In [ ]:
trainingSummary = lrModel.summary

In [ ]:
trainingSummary.accuracy

Out[417]: 0.969758064516129

In [ ]:
from pyspark.ml.evaluation import BinaryClassificationEvaluator

# Evaluate model
evaluator = BinaryClassificationEvaluator()
evaluator.evaluate(predictions)

Out[418]: 0.9951229855810009

In [ ]:
evaluator.getMetricName()

Out[419]: 'areaUnderROC'

In [ ]:
evaluator = BinaryClassificationEvaluator(labelCol="label", rawPredictionCol="prediction")

In [ ]:
evaluator.evaluate(predictions)

Out[421]: 0.9607718405428328

In [ ]:
predictionAndTarget = predictions.select(['label', 'prediction'])

In [ ]:
auc = evaluator.evaluate(predictionAndTarget)

In [ ]:
auc

Out[424]: 0.9607718405428328

In [ ]:
#from pyspark.mllib.evaluation import BinaryClassificationMetrics
#metrics = BinaryClassificationMetrics(predictions)

In [ ]:
# Compute raw scores on the test set
#predictionAndLabels = testData.rdd.map(lambda lp: (float(lrModel.predict(lp.features)), lp.label))

# Instantiate metrics object
#metrics = BinaryClassificationMetrics(predictionAndLabels)

# Area under precision-recall curve
#print("Area under PR = %s" % metrics.areaUnderPR)

# Area under ROC curve
#print("Area under ROC = %s" % metrics.areaUnderROC)

In [ ]:
from pyspark.mllib.evaluation import BinaryClassificationMetrics

results = predictions.select(['prediction', 'label'])
predictionAndLabels=results.rdd
metrics = BinaryClassificationMetrics(predictionAndLabels)

/databricks/spark/python/pyspark/sql/context.py:134: FutureWarning: Deprecated in 3.0.0. Use SparkSession.builder.getOrCreate() instead.
  warnings.warn(


In [ ]:
metrics.areaUnderROC

Out[428]: 0.9607718405428328

In [ ]:
import numpy as np
from sklearn.metrics import confusion_matrix
from pyspark.mllib.evaluation import BinaryClassificationMetrics

def getMetrics(predictions):
  results = predictions.select(['prediction', 'label'])
  predictionAndLabels=results.rdd
  metrics = BinaryClassificationMetrics(predictionAndLabels)
  return metrics

def getConfusion_matrix(predictions):
  y_true = predictions.select("label")
  y_true = y_true.toPandas()
  y_pred = predictions.select("prediction")
  y_pred = y_pred.toPandas()
  cnf_matrix = confusion_matrix(y_true, y_pred)
  return cnf_matrix

def getInfo(predictions):
  arr = getConfusion_matrix(predictions)
  
  FPR = arr[1][0]/sum(arr[1]) # False Positive Rate
  FNR = arr[0][1]/sum(arr[0]) # False Negative Rate
  PPV = arr[0][0]/np.sum(arr, axis=0)[0] #Precision
  TPR = arr[0][0]/np.sum(arr, axis=1)[0] # recall
  ACC = sum(np.diag(arr))/np.sum(arr) # Accuracy
  AUC = getMetrics(predictions).areaUnderROC # AUC
  
  print('+'*100)
  print('\tFalse Positive Rate: ', FPR)
  print('\tFalse Negative Rate: ', FNR)
  print('\n\tPrecision: ', PPV)
  print('\tRecall: ', TPR)
  print('\tAccuracy: ', ACC)
  print('\n\tAUC: ', AUC)
  print('+'*100)

In [ ]:
getInfo(predictions)

/databricks/spark/python/pyspark/sql/context.py:134: FutureWarning: Deprecated in 3.0.0. Use SparkSession.builder.getOrCreate() instead.
  warnings.warn(
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
	False Positive Rate:  0.05555555555555555
	False Negative Rate:  0.022900763358778626

	Precision:  0.9696969696969697
	Recall:  0.9770992366412213
	Accuracy:  0.9655172413793104

	AUC:  0.9607718405428328
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++


## DecisionTree

In [ ]:
from pyspark.ml.classification import DecisionTreeClassifier
dt = DecisionTreeClassifier(labelCol="label", featuresCol="features") #, maxDepth=10
dtModel = dt.fit(trainingData)
predictions_dt = dtModel.transform(testData)

In [ ]:
getInfo(predictions_dt)

/databricks/spark/python/pyspark/sql/context.py:134: FutureWarning: Deprecated in 3.0.0. Use SparkSession.builder.getOrCreate() instead.
  warnings.warn(
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
	False Positive Rate:  0.05555555555555555
	False Negative Rate:  0.030534351145038167

	Precision:  0.9694656488549618
	Recall:  0.9694656488549618
	Accuracy:  0.9605911330049262

	AUC:  0.956955046649703
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++


In [ ]:
# Gets the value of maxDepth or its default value.
dtModel.getMaxDepth()

Out[433]: 5

## Random forest

In [ ]:
from pyspark.ml.classification import RandomForestClassifier
rf = RandomForestClassifier(labelCol="label", featuresCol="features") #, numTrees=10
rfModel = rf.fit(trainingData)
predictions_rf = rfModel.transform(testData)

In [ ]:
getInfo(predictions_rf)

/databricks/spark/python/pyspark/sql/context.py:134: FutureWarning: Deprecated in 3.0.0. Use SparkSession.builder.getOrCreate() instead.
  warnings.warn(
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
	False Positive Rate:  0.0
	False Negative Rate:  0.030534351145038167

	Precision:  1.0
	Recall:  0.9694656488549618
	Accuracy:  0.9802955665024631

	AUC:  0.9847328244274809
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++


In [ ]:
# Gets the value of numTrees or its default value.
rfModel.getNumTrees

Out[436]: 20

> Mô hình **Random Forest** cho AUC lớn nhất là **0.98**